### PR TITLE
User friendly load balancer names

### LINKED ISSUE #29789: AWS: Friendly ELB Names
`a44e18e4c552b11e683bb02fff13e176` is not a very friendly ELB name. The AWS console does not allow for searching/sorting by tags for every resource. We use something like this:

```
[environment]-[app]
staging-my-app
```

It would be nice if we could pick a template to use for naming. A combination of cluster name and pod name might work well.

### PR DESCRIPTION
**Pull request by lpabon**
**What this PR does / why we need it**:
Users have a hard time finding load balancers on their cloud provider consoles due to the non-human friendly name. This fix gives the users two options to avoid this issue. One is to provide a section of the name used for the load balancer, another is to generate it using information from the service object.

**Which issue this PR fixes**: fixes #29789

**Release note**:
```release-note
* New service annotations allow the user to specify the name of the cloud provider load balancer

"kubernetes.io/loadbalancer-name" is a new annotation used on the service
to specify a portion of the name of the cloud provider load balancer.
The maximum length for this portion can be 16 characters. The full
load balancer name will include be created using the following
template: `k8s-{loadbalancer-name:16}-{uid:10}`
The cloudprovider will use this as a hint for inclusion in the name of the
created load balancer, ideally it will be created using the template above.
For example: `"kubernetes.io/loadbalancer-name": "myserver"`, would create
a load balancer with the name: `k8s-myserver-abcgef1234`

"kubernetes.io/loadbalancer-name-auto-generate" is a new annotation on the
service to automatically create a load balancer name based on the name,
namespace, and uuid of the service. The cloud provider loadbalancer name
will be created using the following template: `k8s-{namespace:8}-{name:8}-{uuid:10}`
The cloudprovider will use this as a hint for inclusion in the name of the
created load balancer, ideally it will be created using the template above.
For example, having the annotation `"kubernetes.io/loadlbalancer-name-auto-generate"`, could create a load balancer with the name `k8s-default-myservic-abcgef1234` for a
service called `myservice` on namespace `default`.
```

### CODE REVIEW

**Comment by lpabon** [comment]
cc @justinsb @jbeda

File: pkg/cloudprovider/cloud.go [hunk 1]
```diff
@@ -58,9 +79,41 @@ type Clusters interface {
...
+
+		ret = namespace + "-" + name
+	}
+
+	if len(ret) > 0 {
+		ret = "k8s-" + ret + "-" + strings.Replace(string(service.UID), "-", "", -1)
+	} else {
+		// Fallback to original way to generate the name
+		//GCE requires that the name of a load balancer starts with a lower case letter.
+		ret = "a" + string(service.UID)
```
Comment (calebamiles) [inline pkg/cloudprovider/cloud.go]: is there a reason we can't use `strings.ToLower(service.UID)` here?

Comment (chancez) [inline pkg/cloudprovider/cloud.go]: UUIDs don't start with a letter.

File: pkg/cloudprovider/cloud.go [hunk 2]
```diff
@@ -26,6 +26,27 @@ import (
 	"k8s.io/kubernetes/pkg/controller"
 )
 
+const (
+
```
Comment (xiang90) [inline pkg/cloudprovider/cloud.go]: remove the extra empty line here to follow go style?

File: pkg/cloudprovider/cloud.go [hunk 3]
```diff
@@ -58,9 +79,41 @@ type Clusters interface {
...
 // providers' name length limits.
 func GetLoadBalancerName(service *v1.Service) string {
-	//GCE requires that the name of a load balancer starts with a lower case letter.
-	ret := "a" + string(service.UID)
-	ret = strings.Replace(ret, "-", "", -1)
+
+	var ret string
+
+	if val, ok := service.Annotations[ServiceAnnotationLoadBalancerName]; ok {
+		// The service load balancer portion name has be explicitly set
```
Comment (xiang90) [inline pkg/cloudprovider/cloud.go]:
```go
if len(var) > 16 {
    val = val[:16]
}
ret = val
```

this is good enough?

File: pkg/cloudprovider/cloud_test.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+		},
+	}
+
+	expected := "aabcdef123456"
+	name := GetLoadBalancerName(s)
+	if name != expected {
+		t.Errorf("name[%s] != %s", name, expected)
+	}
+
+	if len(name) > 32 {
```
Comment (xiang90) [inline pkg/cloudprovider/cloud_test.go]: do we need to check this when name already equals to expected?

File: pkg/cloudprovider/cloud_test.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+		t.Error("name length is larger than 32 characters")
+	}
+}
+
+func TestGetLoadBalancerNameAnnotationLoadBalancerName(t *testing.T) {
+
```
Comment (xiang90) [inline pkg/cloudprovider/cloud_test.go]: remove empty line?

File: pkg/cloudprovider/cloud_test.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+			Namespace: "test-namespace",
+			UID:       "abc-def-123-456",
+			Annotations: map[string]string{
+				ServiceAnnotationLoadBalancerName: "",
+			},
+		},
+	}
+
+	name := GetLoadBalancerName(s)
+	if name != "aabcdef123456" {
```
Comment (xiang90) [inline pkg/cloudprovider/cloud_test.go]:
```go
expected := "aabcdef123456"
```

File: pkg/cloudprovider/cloud_test.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	if name != expected {
+		t.Errorf("name[%s] != %s", name, expected)
+	}
+	if len(name) > 32 {
+		t.Error("name length is larger than 32 characters")
+	}
+}
+
+func TestGetLoadBalancerNameAnnotationLoadBalancerAutoGeneratedName(t *testing.T) {
+
```
Comment (xiang90) [inline pkg/cloudprovider/cloud_test.go]: remove the empty line?

File: pkg/cloudprovider/cloud.go [hunk 8]
```diff
@@ -58,9 +79,41 @@ type Clusters interface {
...
+		namespace := service.GetNamespace()
+
+		if len(name) > 8 {
+			name = name[:8]
+		}
+		if len(namespace) > 8 {
+			namespace = namespace[:8]
+		}
+
+		ret = namespace + "-" + name
```
Comment (dims) [inline pkg/cloudprovider/cloud.go]: would `ret = fmt.Sprintf("%.8s-%.8s", name, namespace)` be simpler?

Comment (lpabon) [inline pkg/cloudprovider/cloud.go]: Nice

File: pkg/cloudprovider/cloud.go (hunk 3 above)
Comment (lpabon) [inline pkg/cloudprovider/cloud.go]: Do you mean if we should do any checking from user input?

Comment (xiang90) [inline pkg/cloudprovider/cloud.go]: there is no need to check if len(val) > 0. if it is len(val)=0, then ret will be empty anyway. assigning empty val to ret is harmless.

Comment (lpabon) [inline pkg/cloudprovider/cloud.go]: Cool

**Comment by slack** [comment]
Bit of spelunking on my part, we should be okay on Azure with this change.

**Comment by justinsb** [comment]
LGTM from an AWS perspective.  One potential problem is that not all characters are valid - on AWS the rules are: "This name must be unique within your set of load balancers for the region, must have a maximum of 32 characters, must contain only alphanumeric characters or hyphens, and cannot begin or end with a hyphen."  We should probably sanitize for AWS at least, but maybe also other clouds have different rules.

**Comment by lpabon** [comment]
Not really alphanumeric for AWS, only English (ASCII) alphanumeric: A-Z, a-z, 0-9

**Comment by justinsb** [comment]
/lgtm

I'm not 100% sold on using `z` as the replacement character but I'm not sure anything else is much better.

I'll let someone from the GCE side approve :-)

**Comment by bowei** [comment]
Can you update the release-note with examples of usage and what the resulting LB names will look like?

**Comment by lpabon** [comment]
@bowei I have updated the release notes in the PR comment.

**Comment by bowei** [comment]
cc: @thockin 

This is actually quite a big change. For example, if you add the annotation to a service spec, most of the cloud providers will orphan
existing resources. I think this probably warrants a design doc and discussion targeting 1.9.

Also, you will need e2e tests to make sure nothing breaks in the controllers, cluster upgrade, etc.

**Comment by lpabon** [comment]
@bowei Hi, could you explain how they orphan the cloud providers resources? Is that a feature or a bug?

**Comment by bowei** [comment]
If you add the annotation to an existing service spec, it will no longer be able to find the old resources. I have to try it out to see...

**Comment by lpabon** [comment]
@bowei Good point. I will check also.

**Comment by lpabon** [comment]
I have finished my testing, and this is what I have found on AWS:

1. Create a service (no annotations).
1. Wait until service creates an ELB and the name is recorded in `status.loadBalancer.ingress` of the service.
1. Edit service, and add one of the annotations described here.
1. The controller goes ahead and creates a _new_ ELB with the name generated. It could check if `status.loadBalancer.ingress` is set and delete the ELB but it does not.
1. Now I have _both_ ELBs available, which one of them is an orphan.

It seems like the solution would be to clean up the older load balancer if a new name has been requested.

What do you guys think:

@bowei @justinsb @xiang90 @slack @jbeda

**Comment by mikedanese** [comment]
I can't approve changes to pkg/cloudprovider/* until @bowei is happy because.

> It seems like the solution would be to clean up the older load balancer if a new name has been requested.

How are you going to detect this?

**Comment by bowei** [comment]
@lpabon if you don't mind, could you take this up in 1.9 with a features doc? There are other future impacts regarding changing the names of the load balancer. For example, some cloud providers need to encode identifying information into the name.

**Comment by lpabon** [comment]
@bowei Cool, i will write a feature doc for 1.9.

cc @justinsb

**Comment by jasantunes** [comment]
This would be a nice feature to have.

File: pkg/cloudprovider/cloud.go [hunk 9]
```diff
@@ -60,13 +84,48 @@ type Clusters interface {
...
+
+	// Sanitize for AWS: Only English ASCII alphanumeric characters or hyphens:
+	// We cannot return an error, so we must fix it here.
+	ret = strings.Map(func(c rune) rune {
+		if 'a' <= c && c <= 'z' || 'A' <= c && c <= 'Z' || '0' <= c && c <= '9' || c == '-' {
+			return c
+		}
+
+		// Replace with an acceptable character
+		return 'z'
```
Comment (feiskyer) [inline pkg/cloudprovider/cloud.go]: will this cause name conflicts on AWS?

**Comment by dims** [comment]
@lpabon @justinsb Is this still on for 1.9? (or do we move it out to 1.10)?

**Comment by ncdc** [comment]
The service's UID is auto-generated by the apiserver and it's not possible to specify it. If you somehow "lose" your load balancer service (pretend your cluster died), it still won't be possible to create a new service that can take over and assume the original cloud provider load balancer name, will it?

**Comment by lpabon** [comment]
@ncdc That is completely correct. I worked on this while I was at CoreOS, but that is the conclusion the community came up with, so I am closing this PR.

**Comment by greg-jaunt** [comment]
Please re-open this PR.  This implementation may be faulty, but the need for this functionality still exists.  If you have a large number of services there is no feature in the AWS console to find a the ELB by name.  You must instead click through every single ELB in turn and scan the tags manually in order to find the load balancer you are looking for.  It would be possible to put some piece of the service name into the ELB name so that the names are at least human readable and filterable by text searches.

**Comment by blakebarnett** [comment]
For reference the alb-ingress-controller does this: https://github.com/coreos/alb-ingress-controller/blob/master/pkg/alb/loadbalancer/loadbalancer.go#L130

**Pull request closed without merging**